data/templates/src/pe/dll/template.c

```diff
@@ -20,82 +19,119 @@ void ExecutePayload(void);
 BOOL WINAPI
 DllMain (HANDLE hDll, DWORD dwReason, LPVOID lpReserved)
 {
-    switch (dwReason)
-    {
-        case DLL_PROCESS_ATTACH:
+	switch (dwReason)
+	{
+		case DLL_PROCESS_ATTACH:
 			ExecutePayload();
-            break;
+			break;
 
-        case DLL_PROCESS_DETACH:
-            // Code to run when the DLL is freed
-            break;
+		case DLL_PROCESS_DETACH:
+			// Code to run when the DLL is freed
+			break;
 
-        case DLL_THREAD_ATTACH:
-            // Code to run when a thread is created during the DLL's lifetime
-            break;
+		case DLL_THREAD_ATTACH:
+			// Code to run when a thread is created during the DLL's lifetime
+			break;
 
-        case DLL_THREAD_DETACH:
-            // Code to run when a thread ends normally.
-            break;
-    }
-    return TRUE;
+		case DLL_THREAD_DETACH:
+			// Code to run when a thread ends normally.
+			break;
+	}
+	return TRUE;
+}
+
+// Use a combination semaphore / event to check if the payload is already running and when it is, don't start a new
+// instance. This is to fix situations where the DLL is loaded multiple times into a host process and prevents the
+// payload from being executed multiple times. An event object is used to determine if the payload is currently running
+// in a child process. The event handle is created by this process (the parent) and configured to be inherited by the
+// child. While the child process is running, the event handle can be successfully opened. When the child process exits,
+// the event handle that was inherited from the parent will be automatically closed and subsequent calls to open it will
+// fail. This indicates that the payload is no longer running and a new instance can be created.
+BOOL Synchronize(void) {
+	BOOL bResult = TRUE;
```

Comment:
Why is this set to TRUE? Shouldn't this be set to `FALSE` initially and then set to `TRUE` only when we know the function succeeds?

Reply:
See the comment below for an explanation of why this was needed. Its fine as it is.